### ISSUE TITLE
RLC: warn when `@MustCallAlias` is written on a return type but no parameter type, and vice versa

### ISSUE THREAD

**Issue by msridhar** [typetools/checker-framework#6376]
I don't think that we have any special handling for this case. Because `@MustCallAlias` is a poly annotation, we do get warnings sometimes (if it is placed in a way that is inconsistent with the general poly rules), but (as is clear from this example) not always.

I agree that it would be good for the RLC to warn if there is an `@MustCallAlias` annotation on a parameter but not a return and vice-versa, because that's the only reasonable way to use the annotation.

_Originally posted by @kelloggm in https://github.com/typetools/checker-framework/pull/6375#discussion_r1430295599_

**Comment by jyoo980** [comment]
I can take this on if it's not too critical for the work being done in the RLC.

It should be a matter of making sure we catch cases like the ones below, right?

Cases to NOT issue a warning:
```java
@MustCallAlias T foo(@MustCallAlias U bar) // no error
T foo(U bar) // no error
```

Cases to issue a warning:
```java
T foo(@MustCallAlias U bar) // @MustCallAlias missing on return type
@MustCallAlias T foo(U bar) // @MustCallAlias missing on param type
```

Can we reduce this to saying: if we see a `@MustCallAlias` in the signature of a method, it should occur on the return type and at least one of the parameter types?

**Comment by msridhar** [comment]
Thanks @jyoo980!  I think your understanding is correct, and there is nothing critical / urgent here.  I would expect we could add this check by overriding `visitMethod` in `MustCallVisitor`, or something like that.

**Comment by kelloggm** [comment]
Thanks @jyoo980 for taking this on!

> Can we reduce this to saying: if we see a @MustCallAlias in the signature of a method, it should occur on the return type and at least one of the parameter types?

There are two additional complications that you'll also need to address:
* this rule should also apply to constructors
* its also ok to have an `@MustCallAlias` annotation on the receiver parameter (assuming there is one on the return type), so naively checking the parameter list won't be sufficient

**Comment by jyoo980** [comment]
Thanks for the additional context.

I had a question about a test case:

```java
  public static @MustCallAlias MustCallAliasMethodReturnAndParam mcaneFactory(@MustCallAlias InputStream is)
      throws Exception {
    return new MustCallAliasMethodReturnAndParam(is, false);
  }
```

For the method above, I would expect the return type to be `@MustCallAlias` (as annotated) and the parameter type to be `@MustCallAlias`. However, I am seeing that the result of

```java
AnnotatedTypeMirror returnType = atypeFactory.getMethodReturnType(tree);
```

Is in fact `@PolyMustCall`, which is confusing to me. Is there a reason for this?

**Comment by kelloggm** [comment]
> Where the tree is the MethodTree of mcaneFactory is in fact @PolyMustCall, which is confusing to me. Is there a reason for this

Yes. Under the hood, the Must Call Checker treats `@MustCallAlias` as an alias of `@PolyMustCall`, because the rules for `@MustCallAlias` and polymorphic qualifiers are very similar from its perspective. In particular, the Must Call type of an `@MustCallAlias` parameter should be treated polymorphically with 1) any other `@MustCallAlias` parameters and 2) the return type. This was an implementation shortcut to avoid needing to write that logic from scratch, but it's worked reasonably well for us so far.

The Resource Leak Checker handles all the `@MustCallAlias`-specific rules, IIRC.

**Comment by jyoo980** [comment]
I see, I've been working in the Must Call Visitor for my current work,
maybe I should put my logic in the RLC.

**Comment by jyoo980** [comment]
@kelloggm Thanks for the explanation. I have one more question:

> its also ok to have an @MustCallAlias annotation on the receiver parameter (assuming there is one on the return type), so naively checking the parameter list won't be sufficient

Could you give me an example of what this means? The [draft PR](https://github.com/typetools/checker-framework/pull/6377) I have open here currently handles constructors and regular method calls, I think.

**Comment by kelloggm** [comment]
> Could you give me an example of what this means? The https://github.com/typetools/checker-framework/pull/6377 I have open here currently handles constructors and regular method calls, I think.

I looked for a test case that shows how this works in the RLC, and we don't appear to have one. So maybe you can safely ignore this situation. I was thinking of something like:
```
@MustCallAlias Socket openSocket(@MustCallAlias SocketContainer this) { 
 // open a socket, store it in an @Owning field of the current class, and then return it
}
```

In theory something like that should be verifiable with the RLC, but as I said I can't find a test case - so probably you don't need to worry about that case. Sorry for the confusion!

**Comment by msridhar** [comment]
I feel like the following method _should_ be verifiable by the RLC:
```java
@MustCallAlias Foo somethingFluent(@MustCallAlias Foo this) {
  [...]
  return this;
}
```
But, I don't think we've ever tested it before, and I agree we can skip testing of it for now.

**Comment by jyoo980** [comment]
It looks like there are cases in the E2E tests for the Checker Framework that _do_ test the `implicit this` parameter, like [here](https://dev.azure.com/checkerframework/checkerframework/_build/results?buildId=14352&view=logs&jobId=2405a50a-5fee-5c81-7950-39b6cddddf09&j=2405a50a-5fee-5c81-7950-39b6cddddf09&t=a54d69ec-72cf-5094-4142-ad6c7687eeef), I'll have to account for the case.

**Comment by jyoo980** [comment]
Relevant test case added in [this PR](https://github.com/typetools/checker-framework/pull/6377/files#diff-77c7ac07a9ed880725105344fde79f18d73b9f03b8bfc80f62e9835982ef795cR12-R22) which is also ready for review.
